### PR TITLE
[Flow] Add pattern to canonicalize consecutive pads

### PR DESCRIPTION
**Pull request by qedawkins**
This allows fusing `pad(pad)` into a single pad with the same padding
value. This is added here due to the inclusion of the Affine dialect and
upstream tensor not depending on Affine.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Canonicalizer.cpp [hunk 1]
```diff
@@ -17,6 +18,82 @@ namespace mlir::iree_compiler::IREE::Flow {
 
 namespace {
 
+/// Folds a chain of `tensor.pad` ops with the same constant padding value.
+///
+/// Example:
+///
+/// ```mlir
+///   %0 = tensor.pad %0 low[0, 1] high[0, 2] {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Canonicalizer.cpp]: typo?

```suggestion
///   %0 = tensor.pad %-1 low[0, 1] high[0, 2] {
```

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Canonicalizer.cpp [hunk 2]
```diff
@@ -17,6 +18,82 @@ namespace mlir::iree_compiler::IREE::Flow {
...
+  using OpRewritePattern<tensor::PadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::PadOp padOp,
+                                PatternRewriter &rewriter) const override {
+    if (padOp.getNofold()) {
+      return failure();
+    }
+    auto producerPad = padOp.getSource().getDefiningOp<tensor::PadOp>();
+    if (!producerPad || producerPad.getNofold())
+      return failure();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Canonicalizer.cpp]: nit: wrap with braces. I think we can notify useful message here, it's quite useful in `-debug`.

```suggestion
    if (!producerPad || producerPad.getNofold()) {
      return rewriter.notifyMatchFailure(padOp, "producer is not a foldable tensor.pad op");
    }
```

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/flow_canonicalize.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/flow_canonicalize.mlir:1]: Can you add few negative tests, like tensor.pad with nofold attribute?

**Review by hanhanW — COMMENTED** [review]
Q: do we move the pattern to GlobalOpt?

**Comment by qedawkins** [comment]
> Q: do we move the pattern to GlobalOpt?

Currently GlobalOpt and Flow are sharing the same "CanonicalizeExt" pass, so this is the place that makes the most sense (other than upstream, this pattern should probably go upstream). To move it to GlobalOpt would require making a GlobalOpt specific canonicalization pass, and this pattern fits well in both places.

**Comment by IanWood1** [comment]
> > Q: do we move the pattern to GlobalOpt?
> 
> Currently GlobalOpt and Flow are sharing the same "CanonicalizeExt" pass, so this is the place that makes the most sense (other than upstream, this pattern should probably go upstream). To move it to GlobalOpt would require making a GlobalOpt specific canonicalization pass, and this pattern fits well in both places.

Would this canonicalization pass be used in the planned dispatch partitioning/preprocessing "phase" https://github.com/iree-org/iree/issues/18099?

**Comment by qedawkins** [comment]
I would assume so, yes. Flow could potentially just use a default canonicalization pass after that split also.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/flow_canonicalize.mlir [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      tensor.yield %pad_value1 : f32
+    } : tensor<4x4xf32> to tensor<7x8xf32>
+  util.return %pad1 : tensor<7x8xf32>
+}
+
+// Verify that folding does not happen if it would drop a nofold attribute
+
+// CHECK-LABEL: util.func public @dont_merge_constant_padding_different_vals
+//       CHECK:   tensor.pad
+//       CHECK:   tensor.pad
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/flow_canonicalize.mlir]: Add new line at end of file.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**